#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/RectangleArena.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/gctronic/e-puck/protos/E-puck.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/obstacles/protos/OilBarrel.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/buildings/protos/SimpleBuilding.proto"

WorldInfo {
  title "Disaster Response Robot Coordination"
  info [
    "Robot swarm coordination for disaster response"
    "Multiple robots bid for tasks and execute missions"
    "Integration with Sei blockchain for payments"
  ]
  basicTimeStep 32
  contactProperties [
    ContactProperties {
      material1 "InteriorWheelMat"
      coulombFriction [
        0, 2, 0
      ]
      frictionRotation -0.9648 0
      bounce 0
      forceDependentSlip [
        10, 0
      ]
    }
    ContactProperties {
      material1 "ExteriorWheelMat"
      coulombFriction [
        0, 2, 0
      ]
      frictionRotation 0.9648 0
      bounce 0
      forceDependentSlip [
        10, 0
      ]
    }
  ]
}
Viewpoint {
  orientation -0.04 0.99 0.12 2.35
  position -15 8 25
  follow "supervisor"
  followSmoothness 0
}
TexturedBackground {
}
TexturedBackgroundLight {
}
RectangleArena {
  translation 0 0 -0.1
  floorSize 20 20
  floorTileSize 2 2
  wallHeight 0.5
  wallAppearance PBRAppearance {
    baseColorMap ImageTexture {
      url [
        "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/default/worlds/textures/red_brick_wall.jpg"
      ]
    }
    roughness 1
    metalness 0
  }
}

# Disaster Zone A (Red)
DEF ZONE_A Transform {
  translation -6 -6 0.01
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 1 0.2 0.2
        transparency 0.7
      }
      geometry Box {
        size 4 4 0.01
      }
    }
  ]
}

# Disaster Zone B (Yellow) 
DEF ZONE_B Transform {
  translation 6 -6 0.01
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 1 1 0.2
        transparency 0.7
      }
      geometry Box {
        size 4 4 0.01
      }
    }
  ]
}

# Disaster Zone C (Blue)
DEF ZONE_C Transform {
  translation 0 6 0.01
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 0.2 0.2 1
        transparency 0.7
      }
      geometry Box {
        size 4 4 0.01
      }
    }
  ]
}

# Obstacles and debris
DEF OBSTACLE_1 OilBarrel {
  translation -3 0 0.44
  rotation 0 0 1 1.5708
  name "obstacle_1"
}
DEF OBSTACLE_2 OilBarrel {
  translation 3 0 0.44
  rotation 0 0 1 0
  name "obstacle_2"
}
DEF OBSTACLE_3 OilBarrel {
  translation 0 -3 0.44
  rotation 0 0 1 3.14159
  name "obstacle_3"
}

# Damaged buildings
DEF BUILDING_1 SimpleBuilding {
  translation -8 8 0
  name "building_1"
  floorHeight 3
  roofHeight 1
}
DEF BUILDING_2 SimpleBuilding {
  translation 8 8 0
  name "building_2"
  floorHeight 3
  roofHeight 1
}

# Robot 1 - UGV Alpha
DEF UGV_ALPHA E-puck {
  translation -2 -2 0
  rotation 0 0 1 0
  name "ugv_alpha"
  controller "ugv_agent"
  supervisor FALSE
  camera_width 320
  camera_height 240
}

# Robot 2 - UGV Beta  
DEF UGV_BETA E-puck {
  translation 2 -2 0
  rotation 0 0 1 1.5708
  name "ugv_beta"
  controller "ugv_agent"
  supervisor FALSE
  camera_width 320
  camera_height 240
}

# Robot 3 - UGV Gamma
DEF UGV_GAMMA E-puck {
  translation 0 -8 0
  rotation 0 0 1 3.14159
  name "ugv_gamma"
  controller "ugv_agent"
  supervisor FALSE
  camera_width 320
  camera_height 240
}

# Supervisor Robot for coordination
DEF SUPERVISOR Robot {
  translation 0 0 5
  name "supervisor"
  controller "coordinator_supervisor"
  supervisor TRUE
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 0.8 0.8 0.8
        transparency 0.8
      }
      geometry Sphere {
        radius 0.1
      }
    }
    Emitter {
      channel 1
    }
    Receiver {
      channel 1
    }
  ]
}